tree/tree/src/InternalTreeUtils.cxx

```diff
@@ -403,6 +408,56 @@ std::vector<std::unique_ptr<TChain>> MakeFriends(const ROOT::TreeUtils::RFriendI
    return friends;
 }
 
+std::vector<std::string> ExpandGlob(TString glob)
+{
+   TString directory;
+
+   // TODO: Also implement this pattern in some other free function
+   // This is practically Python's os.path.split
+   Int_t slashpos = glob.Last('/');
+   if (slashpos >= 0) {
+      directory = glob(0, slashpos); // Copy the directory name
+      glob.Remove(0, slashpos + 1);  // and remove it from glob
+   } else {
+      directory = gSystem->UnixPathName(gSystem->WorkingDirectory());
+   }
+
+   const char *epath = gSystem->ExpandPathName(directory.Data());
+   void *dir = gSystem->OpenDirectory(epath);
+   delete[] epath;
+
+   if (dir) {
+      // create a TList to store the file names (not yet sorted)
+      TList l;
+      TRegexp re(glob, true);
+      const char *file;
+      TString s;
+      while ((file = gSystem->GetDirEntry(dir))) {
+         if (!strcmp(file, ".") || !strcmp(file, ".."))
+            continue;
+         s = file;
+         if ((glob != file) && s.Index(re) == kNPOS)
+            continue;
+         // Using "/" as separator here as it was done in TChain::Add
+         // In principle this should be using the appropriate platform separator
+         l.Add(new TObjString((directory + "/" + file).Data()));
+      }
+      gSystem->FreeDirectory(dir);
+      // sort the files in alphanumeric order
+      l.Sort();
+
+      // Convert TList<TString> to std::vector<std::string>
+      std::vector<std::string> ret;
```

Comment:
The disadvantage of this interface (creating `TObjString` then copying to `std::string` rather than just returning the the `TList`) is that this increases the memory churns (where the gain is 'solely' a more modern looking interface).

Reply:
Yeah, I am also undecided. If I keep the `TList`, then I need to document that and the caller will need to do things a bit more clunkily:
```cpp
const auto files = ExpandGlob(files);
for (const auto *objstr: ROOT::RangeStaticCast<const TObjString *>(files)){
    std::cout << objstr->GetName() << std::endl;
}
```
So I see that as a greater disadvantage at this point, but I'm a bit on the fence about it

Reply:
I'm not sure what Philippe refers to exactly with "memory churn", the conversion will not take more than a millisecond and `ExpandGlob` is not called in a hot loop, so I'm happy to take the "hit" for a more usable interface.

Reply:
memory churns:  Each element of the list is created once with 2 allocations (TObjString and its TString)) and then the copy in the std::vector means that this is copied once more into the std::string.  Indeed, "not too serious" but still ...

> . If I keep the TList, then I need to document that and the caller will need to do things a bit more clunkily:

Actually you can solves "that" problem by using:
```
std::unique_ptr<TList*>  ExpandGlob(TString glob);  // and call SetOwner(kTRUE) inside.
```

Reply:
the extra copy for 10k(!) non-small strings takes 0.8 milliseconds on my laptop

Reply:
In that case why would you worry about the copy of the argument? :) :)

More seriously, that still leave the interface to be 'inconsistent' (taking a `TString` and returning `std::string's).

If the copy of the return value does not cost anything, we may as well copy the argument too:
```
std::vector<std::string> ExpandGlob(std::string glob);
```

Reply:
In a previous comment I was saying that using `TString` as input parameter has the "added benefit" that I can then create in the future another overload of this function that takes `std::string` but internally uses only C++ standard library classes (except maybe just for `TRegexp` since there is no better standard alternative).

But, we could just simply agree that this function is only for `TChain` and classes that want to replicate that behaviour. In the future, i.e. for RNTuple and ROOT 7, we can still move to another function with more modern implementation. 

It's an internal function anyway, and I believe it should stay that way, so we will be able to change name or overload it as needed when the need comes.

Reply:
I am thinking that for this PR to reduce code churn (i.e. use the code of the function 'as-is') and still have a consistent interface we should use either:
```
std::unique_ptr<TList*>  ExpandGlob(TString glob);
```
or
```
std::vector<std::string> ExpandGlob(std::string glob);  // First line of the function is TString glob = in_glob;
```

Reply:
In the latest status of the commit I changed to a less unexpected function signature, which shouldn't need specific extra comments. Let me know what you think